Implement OpenStack provider for openshift-ansible

lhuard1A opened the pull request:
Implement OpenStack support in openshift-ansible

@detiber @twiest @jwhonce : This PR still needs some polishing before being mergeable, but I wanted to let you know that I have started working on OpenStack deployment.

detiber commented:
@lhuard1A One initial comment I have is that I see you are using the new openstack.py inventory provider, which uses the shade library (which hasn't yet been packaged for fedora/rhel).

Ideally, any dependencies we add to this repo would be available in either fedora or rhel (with the idea that we'd do the rhel packaging if a fedora package is present).

Would it be possible to try to get this working using the nova.py inventory until shade has been packaged?

lhuard1A commented:
Hello,

I have now reached a point where the OpenStack provider for openshift-ansible is working for me.
@detiber : Could you please review it and share feedback? ;-)

I have reverted to the `nova.py` dynamic inventory to remove the dependency to `shade`.

In order to check that all the dependencies introduced are present on Fedora, I tested the OpenStack provider from a Fedora based docker image: https://registry.hub.docker.com/u/lhuard1a/openshift-ansible/dockerfile/

detiber commented on playbooks/os/openshift-cluster/launch.yml:
```diff
@@ -0,0 +1,35 @@
+  tasks:
+  - fail:
+      msg: "Deployment type not supported for OpenStack provider yet"
+    when: deployment_type == 'online'
```
This should probably be changed to:
```
when: deployment_type in ['online', 'enterprise']
```
Since enterprise deployments will not currently work (because of lack of registration support).

lhuard1A replied:
I would like to use pre-configured RHEL 7.1 VM image.
I need to pre-configure them because openshift-ansible doesn’t do the registration yet.
But still, I need to use the `enterprise` deployment type, otherwise, the `origin` deployment type will setup yum repositories which are not suitable for RHEL.

detiber commented on playbooks/os/openshift-cluster/config.yml:
```diff
@@ -0,0 +1,39 @@
+---
+# TODO: need to figure out a plan for setting hostname, currently the default
+# is localhost, so no hostname value (or public_hostname) value is getting
+# assigned
```
Is the metadata not being returned properly or lacking the hostname values?  openshift_facts should be setting the hostname and public_hostname variables based on the instance metadata.  It should only fall back to hostname -f if the metadata does not provide the correct values.

lhuard1A replied:
@detiber:
Here is the issue that I have now with my OpenStack setup:

When the VMs are created, their hostname is initially the symbolic name (eg.: lenaic-master-1c577.novalocal) (passed by OpenStack to the VM via the metadata)

##### If I put `openshift_hostname: "{{ ansible_default_ipv4.address }}"` in `playbooks/openstack/openshift-cluster/config.yml` (similar to what is done for GCE)

The `Set hostname` task of `roles/openshift_common/tasks/main.yml` changes the hostname to use the IP (eg.: 192.168.211.2).
The result is that all machines have, as hostname, 192, and all machines have a different domain, (eg.: 168.211.2)
If the machines are rebooted, they get their symbolic name (lenaic-master-1c577.novalocal) as hostname again.

##### If I put `openshift_public_hostname: "{{ ansible_default_ipv4.address }}"` in `playbooks/openstack/openshift-cluster/config.yml` (similar to what is done for AWS)

The machines are correctly named with their symbolic name. But the machines are referenced by name from other and because there has been no DNS registration, the name resolution fails

```
[openshift@lenaic-master-f1d2d ~]$ osc get nodes
NAME                          LABELS        STATUS
lenaic-node-64c82.novalocal   Schedulable   <none>    NotReady
lenaic-node-a5fc9.novalocal   Schedulable   <none>    NotReady
```
* hostname of the master is correct (in the prompt)
* name of the nodes are correct
* The nodes are not ready because DNS resolution doesn’t work.

Ex of logs:
```
Jun 02 04:10:54 lenaic-master-f1d2d.novalocal openshift-sdn-master[12945]: E0602 04:10:54.498659 12945 controller.go:129] Failed to lookup IP address for minion lenaic-node-64c82.novalocal.
```

##### If I set neither `openshift_hostname` nor `openshift_public_hostname` in `playbooks/openstack/openshift-cluster/config.yml` (similar to what is done for libvirt)

Same issue as above: the hostname of the machines is correctly the symbolic name bus as the DNS resolution doesn’t work, symbolic names cannot be used on the master

```
[openshift@lenaic-master-5947c ~]$ osc get nodes
NAME                          LABELS        STATUS
lenaic-node-13248.novalocal   Schedulable   <none>    NotReady
lenaic-node-580e1.novalocal   Schedulable   <none>    NotReady
```

Ex of logs:
```
Jun 02 04:51:28 lenaic-master-5947c.novalocal openshift-sdn-master[12928]: E0602 04:51:28.281472 12928 controller.go:129] Failed to lookup IP address for minion lenaic-node-13248.novalocal.
```

##### Conclusion

I wish I could put neither `openshift_hostname` nor `openshift_public_hostname` in `config.yml` because the machines already know their names thanks to metadata.
The issue I have with that is that machines reference each other via DNS names and DNS resolution doesn’t work.
Ex:
```
ps:
/usr/bin/openshift-sdn -minion -etcd-endpoints=https://lenaic-master-5947c.novalocal:4001
logs:
… lenaic-node-13248.novalocal openshift-sdn-node[12972]: … 501: All the given peers are not reachable 
```

Putting neither `openshift_hostname` nor `openshift_public_hostname` works for the libvirt setup because libvirt uses a dnsmasq to manage the VMs’ network and that dnsmasq is used as both DHCP and DNS server and dnsmasq’s DHCP leases are published in dnsmasq’s DNS.

So, for OpenStack, I would like to put IP everywhere, except for the hostnames.

detiber commented on playbooks/os/openshift-cluster/vars.yml:
```diff
@@ -0,0 +1,33 @@
+    image:
+      url:
+      name: rhel-7.1
+      sha256:
```
It looks like the url and sha256 params are never used.

detiber commented on playbooks/os/openshift-cluster/vars.yml:
```diff
@@ -0,0 +1,33 @@
+  - 8.8.4.4
+openstack_ssh_keypair: "{{ lookup('env', 'LOGNAME') }}_key"
+openstack_ssh_public_key: "{{ lookup('file','~/.ssh/id_rsa.pub') }}"
+openstack_flavor_ram: 2048
```
We should probably separate out the flavor configs so that they can be specified differently for masters and nodes.

Also, it would be nice to specify flavor by other attributes other than ram (at least id or name).

detiber commented on playbooks/os/openshift-cluster/vars.yml:
```diff
@@ -0,0 +1,33 @@
+---
+openstack_network_prefix: 'openshift-ansible-{{ cluster_id }}'
```
These variables should be override-able using env variables.  It's not a big deal for the libvirt provider, since it is used  for local instance creation, but for providers that can be used to host production instances, we should allow more flexibility to the user without having to overwrite the vars files.

detiber commented:
@lhuard1A I am starting to get excited to try this out.  Outside of the comments I made on the diffs, We should have a README associated with the openstack provider. It'll also need to list the requirements needed to run this provider (namely that the openstack instance has neutron and heat enabled).

twiest commented on bin/cluster:
```diff
@@ -172,7 +174,7 @@ if __name__ == '__main__':
     cluster = Cluster()
 
-    providers = ['gce', 'aws', 'libvirt']
+    providers = ['gce', 'aws', 'libvirt', 'os']
```
Hmm, I'm afraid that 'os' is just too generic. That could be confused for meaning bare metal (what we today call 'byo'), or OpenShift (which wouldn't make sense, of course).

Could we just call this 'openstack' ?

Or is there another acronym that openstack uses, maybe ostack or something?

lhuard1A commented:
@detiber  @twiest
I have renamed `os` to `openstack`,
I have made all the variables of `vars.yml` modifiable thanks to the `--option` parameter of `bin/cluster`
I added a README.
Could you have a look?

twiest commented:
Looks good to me.

@detiber Let me know if / when I should merge this please.

lhuard1A commented:
Jenkins’ pylint is failing because of the `nova.py` dynamic inventory script that I just blindly copied from https://github.com/ansible/ansible/blob/devel/plugins/inventory/nova.py

Should we modify the upstream script to make it comply with our policy or should we update https://github.com/openshift/openshift-ansible/blob/master/docs/best_practices_guide.adoc to allow disabling pylint globally for files copied from outside?

twiest commented:
Hey @lhuard1A, we discussed this in the arch team meeting and we created a card to add an "exclude" list to pylint. We'll add all upstream files (inventories) to that exclude list once it has been added.

For now, please add a full file disable for this inventory script:
http://stackoverflow.com/questions/10238473/pylint-disable-all-warnings-for-a-file

Please don't waste any time modifying the upstream code to comply with our pylint policy as we'll want to be able to replace this file with the latest upstream version. We already know that upstream has no interest in complying with pylint, so they likely won't take any pylint only changes anyway.

Thanks!

detiber commented:
@twiest I'm good with the state of this once the merge conflicts are handled.

lhuard1A commented:
@twiest : PR rebased and squashed

twiest commented:
@lhuard1A Another great PR Lenaic, thanks for submitting it! :)    :+1:

newgoliath commented:
I'm unclear where this should be run from:

```
[root@cntl1 openshift-ansible(openstack_admin)]# ./deploy-images.sh
ERROR: Inventory script (inventory/openstack/hosts/nova.py) had an
execution error: Unable to find configfile in
/root/openshift-ansible/nova.ini, /root/nova.ini, /etc/ansible/nova.ini
```

On Mon, Jun 8, 2015 at 10:57 AM, Thomas Wiest <notifications@github.com>
wrote:

> Merged #186 <https://github.com/openshift/openshift-ansible/pull/186>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/openshift-ansible/pull/186#event-325339819>.
>



-- 
Judd Maltin
T: 917-882-1270
Of Life immense in passion, pulse, and power,  Cheerful—for freest action
form’d, under the laws divine,  The Modern Man I sing. -Walt Whitman

detiber commented:
@newgoliath the easiest place to run from is the root directory of the openshift-ansible checkout (which it looks like you are doing).

The error is pointing towards trying to use the openstack provider without a nova.ini file configured somewhere in the search path for the openstack dynamic inventory.  https://github.com/ansible/ansible/blob/devel/plugins/inventory/nova.ini can be used as a template for this file.
